syntax="proto3";
package GameProto;

message Position{
	float x=1;
	float y=2;
	float z=3;
}
//S2C
enum ServerEventCode {
	LogInSuccess = 0;
	RegisterSuccess = 1;
	EnterRoomSuccess = 2;
	
	LogInError_AccountDontExist = 3;
	LogInError_PasswordWrong = 4;
	LogInError_ReLogIn = 5;

	RegisterError_AccountAlreadyExist = 6;
	
	EnterRoomError_DontLogIn = 7;
	EnterRoomError_RoomIsFull = 8;

	InternalError = 9;

    BroadRoomInfo = 10;
	JumpToBattleServer = 11;
	S2CSync = 12;
}

//C2S
enum ClientEventCode {
	LogIn = 0;
	Regist = 1;
	EnterRoom = 2;
	C2SSync = 3;
	Follow = 4;
}

//玩家状态
message PlayerState{
	int32 fid=1;
	string name=2;
	float yaw=3;
	Position pos=4;
	int32 state=5;
}

//玩家一次操作
message PlayerInput{
	string name=1;
	float torque=2;
	bool grasp=3;
	bool attach=4;
}

message RoomInfo {
    int32 id = 1;
    int32 maxusers = 2;
    repeated string members = 3;
}

//客户端数据包
//登录：type=LogIn,name,password
//注册：type=Regist,name,password
//进房间：type=EnterRoom,name,id=room_id
//同步：type=Sync,input
//追帧：type=Follow,name,id=fid
message ClientMsg{
	ClientEventCode type=1;
	string name=2;
	string password=3;
	int32 id=4;
	PlayerInput input=5;
}


//服务器数据包
//登录成功：type = LogInSuccess
//注册成功：type = RegistSuccess
//进房间成功：type = EnterRoomSuccess,fid=房间人数
//登录，注册，进房间失败：type=Failure,str
//跳转服务器: type=JumpToBattleServer,str=ip:port
//同步：type = Sync,fid,inputs,(states?)
message ServerMsg{
	ServerEventCode type=1;
	int32 fid=2;
	string str=3;
	repeated PlayerInput inputs=4;
	repeated PlayerState states=5;
    repeated RoomInfo roominfos=6;
}
